### PR TITLE
Compression built around flate package

### PR DESCRIPTION
**Pull request by ligser**
Simple reader and writer that support deflate compression.
Can be used as the wrapper around the wsutil.Writer and wsutil.Reader structs.

Code highly inspired by Gorilla's WebSocket approach, but I not properly sure about pools for flate.Reader/flate.Writer.

### CODE REVIEW

**Comment by ligser** [comment]
Whoops. Just realize that the writer will not work without proper Header with the RSV1 bit set to 1.
Rework that later.

File: wsutil/compress.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	flateReader io.ReadCloser
+}
+
+func NewCompressReader(r io.Reader) CompressorReader {
+	return &compressReader{
+		flateReader: flate.NewReader(
+			io.MultiReader(
+				r,
+				bytes.NewReader(deflateFinal[:]),
+				bytes.NewReader(tail[:]),
```
Comment (cristaloleg) [inline wsutil/compress.go]: `[:]` is redundant 'cause `tail` is already a slice 😉 
or maybe it's better to convert `tail` to an array as a `defalateFinal`

File: wsutil/compress.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func (cw *compressWriter) Write(p []byte) (n int, err error) {
+	if cw.flateWriter == nil {
+		return 0, ErrWriteClose
+	}
+
+	n, err = cw.flateWriter.Write(p)
+	if err != nil {
+		return
```
Comment (cristaloleg) [inline wsutil/compress.go]: Maybe a `return 0, nil` ? 😄 (the same below)

File: wsutil/compress.go (hunk 1 above)
Comment (ligser) [reply wsutil/compress.go]: Yes, the second option sounds good :)

File: wsutil/compress.go (hunk 2 above)
Comment (ligser) [reply wsutil/compress.go]: Hmm. Do you think hide error from the flateWriter is a good idea?

**Comment by ligser** [comment]
Try to implement autobahn testsuite pass with compression enabled and it dont work :)
I’ll try to fix it and update PR

**Comment by ligser** [comment]
@gobwas complete work with autobahn test suite. Now it all green. Travis does not green because full test suite with compression enabled take almost 10 minutes to fully pass.
The final solution deeper integrated into the codebase, that not just the wrapper around it. That needed because of actually compression is invasive. It changes many things like RSV, frame length (and framing at all), require proper «tails» and strip it out also.
I don't like that writer struct now more stateful, it requires intermediate states like RSV1 bit state in fragmented messages, but I see the same approach for opCode and State and think it not as bad as it looks like.
One more disadvantage is Reader implementation. It has the buffer in the Decompressor that can growth uncontrollable. I think there is also be a more clean solution for that but have no idea how it can be achieved. Maybe I should reserve a fixed size buffer and limit uncompressed data to its size because streaming decompression is not possible. Does not sounds like a reliable solution too.

If you have an idea how to do the same thing better — you can tell me and I'll try to implement it.

**Comment by ligser** [comment]
That's me again :) 

Rework compression extension again to produce less complex integration. Now it actually can be a plugin in the separate repo if you prefer that way :)
Aside, this also means that from now wsutil not support compression at all via high-level API.
It can be integrated into wsutil later in a separate request. It requires some more thinking.

**Comment by gobwas** [comment]
Hi @ligser, looks like you really simplified the things.
Let me review it in details and then we could decide what to do with it (separate package or some integration in `wsutil`).

Have a nice New Year holidays! 🎆

**Comment by navossoc** [comment]
@ligser are you using this code anywhere?
Looks promising, I want to give it a try.

[]'s

**Comment by ligser** [comment]
@navossoc yes, I use compression extension more than a year in two modes:
1. Broadcast — WS-server with a pub-sub model when I deliver the same message for many clients.
2. Request-Response — WS-server for realtime client-server communication.

Both servers can handle thousands simultaneously connected clients with low memory usage (about 300Mb for 10000 clients) on my setup.

I do not try to deploy that code to the large servers. Instead, I use tens of medium-sized VMs and a good load-distributor.

**Comment by navossoc** [comment]
I see, I'm planing to use it on ~2k realtime connections, memory will not be an issue.
At the moment I'm more concerned about the cpu usage.

Anyway, just testing to see the effects ;)

I need to do some house keeping on my code first, then I'll try enabling the compression.

However, as you said you are already using on a large scale, it reassures me a lot.

Cheers.

**Comment by ligser** [comment]
@navossoc try to do some load-testing before. jMeter a good tool to do that. And try to reuse memory with something like a sync.Pool or https://github.com/gobwas/pool.
I also implement a PreparedMessage (like in [gorilla](https://github.com/gorilla/websocket/blob/master/prepared.go)) concept for a broadcaster to reuse memory as much as it possible.

**Comment by navossoc** [comment]
I had used it in the initial tests, it works well, but sometimes it was a bit clumsy, some connections didn't complete, etc ...
I don't have the concept of broadcast, since the messages are very unique.

I may need to think about on how to decrease the number of goroutines, so far is 2 per connection.

**Comment by gobwas** [comment]
Hi @ligser,

Thank you for a lot of effort you've put on this PR.
However, I'm going to close it in connection with this [implementation](#118). In short, I don't want to bring any specific knowledge of any negotiated extension in `ws` or `wsutil` packages code.

Since you have lot of experience of using permessage-deflate, I would be glad to see your review/comments on things from the referenced PR.

**Comment by ligser** [comment]
Hi! Good to see that you found a time to implement a compression extension natively into the ws library!

I'll take a look more precisely, but at a glance proposed API looks good! Also, the goal to not couple work with negotiated extensions into library code is the right thing to do. That's «extensions» after all :)

**Comment by gobwas** [comment]
Yeah, exactly! Thanks again. 👍

**Pull request closed without merging**